runtime/src/main/java/org/corfudb/runtime/view/stream/StreamAddressSpace.java

```diff
@@ -0,0 +1,288 @@
+package org.corfudb.runtime.view.stream;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Function;
+
+import lombok.Synchronized;
+
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.StreamOptions;
+
+/**
+ * This is an abstract class which represents the space of addresses of a stream. It is based on the
+ * assumption that the address space of a stream can be traversed forward and backwards.
+ *
+ * The abstract method 'findAddresses' defines how to traverse the space of unexplored addresses,
+ * i.e, addresses between the latest synced point (oldTail) and the latest update (newTail).
+ * For example, the implementation of BackpointerAddressSpace assumes the use of backpointers
+ * to determine the previous log entry for a stream.
+ *
+ * Created by amartinezman on 4/23/18.
+ */
+public abstract class StreamAddressSpace implements IStreamAddressSpace {
+
+    protected CopyOnWriteArrayList<Long> addresses;
```

Comment:
Why is this needed?

Reply:
Moreover, shouldn't a collection with range query abilities (say, SkipList) be used?

Reply:
Also, it seems like inserts are expensive on this container type (i.e. O(n) inserts):

List                 | Add  | Remove | Get  | Contains | Next | Data Structure
---------------------|------|--------|------|----------|------|---------------
ArrayList            | O(1) |  O(n)  | O(1) |   O(n)   | O(1) | Array
LinkedList           | O(1) |  O(1)  | O(n) |   O(n)   | O(1) | Linked List
CopyOnWriteArrayList | O(n) |  O(n)  | O(1) |   O(n)   | O(1) | Array